From 2d88709059f0adc5bb5a1fcb00bead41fb095880 Mon Sep 17 00:00:00 2001
From: Aihua Xu <aihuaxu@apache.org>
Date: Fri, 5 Feb 2016 09:07:34 -0500
Subject: [PATCH 0486/1164] CDH-36946 HIVE-12790: Metastore connection leaks
 in HiveServer2 (Naveen Gangam via Aihua Xu,
 reviewed by Thejas M Nair, Aihua Xu)

Conflicts:
	ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java

Change-Id: Ib75e1edecbf5402091dc54d92a36219121c15fab
---
 .../org/apache/hadoop/hive/ql/exec/TaskRunner.java |   12 ++++++++++++
 .../hadoop/hive/ql/session/SessionState.java       |   11 ++++++++---
 2 files changed, 20 insertions(+), 3 deletions(-)

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/exec/TaskRunner.java b/ql/src/java/org/apache/hadoop/hive/ql/exec/TaskRunner.java
index 94ce0d0..379ab27 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/exec/TaskRunner.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/exec/TaskRunner.java
@@ -21,8 +21,11 @@
 import java.io.Serializable;
 import java.util.concurrent.atomic.AtomicLong;
 
+import org.apache.hadoop.hive.ql.metadata.Hive;
 import org.apache.hadoop.hive.ql.session.OperationLog;
 import org.apache.hadoop.hive.ql.session.SessionState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * TaskRunner implementation.
@@ -44,6 +47,8 @@ protected Long initialValue() {
 
   protected Thread runner;
 
+  private static transient final Logger LOG = LoggerFactory.getLogger(TaskRunner.class);
+
   public TaskRunner(Task<? extends Serializable> tsk, TaskResult result) {
     this.tsk = tsk;
     this.result = result;
@@ -74,6 +79,13 @@ public void run() {
       SessionState.start(ss);
       runSequential();
     } finally {
+      try {
+        // Call Hive.closeCurrent() that closes the HMS connection, causes
+        // HMS connection leaks otherwise.
+        Hive.closeCurrent();
+      } catch (Exception e) {
+        LOG.warn("Exception closing Metastore connection:" + e.getMessage());
+      }
       runner = null;
     }
   }
diff --git a/ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java b/ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java
index 36f9c60..5ad0499 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/session/SessionState.java
@@ -1374,9 +1374,14 @@ public void close() throws IOException {
       tezSessionState = null;
     }
 
-    closeSparkSession();
-    registry.closeCUDFLoaders();
-    dropSessionPaths(conf);
+    try {
+      closeSparkSession();
+      registry.closeCUDFLoaders();
+      dropSessionPaths(conf);
+    } finally {
+      // removes the threadlocal variables, closes underlying HMS connection
+      Hive.closeCurrent();
+    }
   }
 
   public void closeSparkSession() {
-- 
1.7.9.5

